### PR TITLE
feat(training): add NVIDIA ARM64 support for LeRobot 0.6.1

### LINKED ISSUE #1446: feat(training): support LeRobot 0.6.1 on NVIDIA ARM64
## Summary

Add NVIDIA ARM64 support for LeRobot 0.6.1 across training, local evaluation, dataset handling, and the development container.

## Acceptance criteria

- LeRobot 0.6.1 dependencies resolve on NVIDIA ARM64 and x86_64.
- Shared-file Parquet and video layouts evaluate correctly.
- Dataset-controlled paths remain contained within the dataset root.
- GPU and Vulkan capabilities are available in the development container.
- Automated tests cover ARM64 compatibility and shared-file behavior.

### PR DESCRIPTION
**Pull request by IoTDan**
# Pull Request

## Description

Adds NVIDIA ARM64 (Jetson Thor / CUDA 13) support so LeRobot imitation-learning work can run on local ARM64 hardware as well as the existing x86_64 AzureML and OSMO targets, and bumps LeRobot to **0.6.1**. The dataset-consuming code paths were updated for LeRobot's newer shared-file layout, and several related correctness fixes travel with the change.

> The decoder choice drives much of this PR. LeRobot's `dataset` extra requires TorchCodec, which publishes no SBSA CUDA 13 wheel, and `get_safe_default_video_backend()` selects TorchCodec whenever it is importable. Keeping TorchCodec on x86_64 only would have decoded training and evaluation frames through different backends per architecture, so the branch standardizes on **PyAV** everywhere.

### Dependencies

- Moved `training/il/lerobot` from `lerobot[dataset]==0.6.0` to `lerobot[av-dep]==0.6.1` and removed `torchcodec` as both a direct dependency and a resolution override.
- Declared `pandas` and `jsonlines` directly, since *lerobot.datasets* imports them eagerly and they shipped in the extra that is no longer selected. Runtime installs use `uv pip install --no-deps`, so both must be in the lock.
- Removed the `wandb` pin. All three LeRobot entry points (*azureml-train-entry.sh*, *lerobot-train-osmo-entry.sh*, and *train.py*) pass `--wandb.enable=false`, and `wandb` only appears in LeRobot's unselected `training` extra.
- Widened `[tool.uv].environments` to cover `x86_64` and `aarch64`, and added the `jetson-ai-lab-cu130` index as an `explicit` source for ARM64 `torch` and `torchvision` only.
- Existing CVE-motivated pins (`gitpython`, `urllib3`, `cryptography`) are unchanged.

### Dev container

- Rebuilt *.devcontainer/Dockerfile* as a two-stage image. The CUDA stage is digest-pinned, installs exact-version CUDA and CUPTI packages, and adds cuDSS plus NVPL BLAS/LAPACK on arm64 only. The final base moved to `devcontainers/base:ubuntu-24.04`.
- Added `hostRequirements.gpu: "optional"` so GPU-less hosts still build, plus the `--group-add` entries CUDA needs. `/dev/nvmap` is owned by `video` and `/dev/dri/renderD*` by `render`; without both, `cudaGetDeviceCount()` fails with error 801 while `nvidia-smi` still looks healthy.
- Reworked the `osmo-cli` install from an x86_64-only skip into an architecture dispatch that installs the checksum-verified OSMO 6.3.0 client on `arm64` as well.

### Data viewer

- **Blob sync** now requests blob metadata and skips ADLS Gen2 directory blobs (`hdi_isfolder`) along with HuggingFace `.cache` entries. Writing a directory blob as a file previously blocked creation of the directory beneath it.
- **Cached episode clips** are validated with a bounded *ffmpeg* probe before being served; invalid clips are deleted and regenerated.
- **TrajectoryEditor** was split into a wrapper and an inner frame component keyed by episode and frame, replacing a synchronization effect that discarded in-progress edits whenever the adjustments map changed for an unrelated frame. Clearing all adjustments now also resets the editor inputs, which the previous version left showing stale deltas.

### Evaluation and training scripts

- *run-local-lerobot-eval.py* reads the `meta/episodes/**.parquet` metadata layout, resolves data and video files through the `data_path` and `video_path` templates, filters rows by `episode_index`, and decodes only each episode's timestamp window. The legacy chunk-scan and `episodes.jsonl` paths remain as fallbacks, and metadata lookups are guarded so datasets lacking the newer columns fall back rather than raising.
- *download_dataset.py* skips ADLS directory-marker blobs instead of trying to download them as files.

### Infrastructure

- *03-deploy-osmo.sh* labels the pre-created `osmo-default-admin` secret, plus `db-secret` and `redis-secret` when they are Key Vault-backed, with `secrets-store.csi.k8s.io/managed=true`. The CSI driver otherwise refuses to adopt secrets it does not own. The change is metadata-only and reads no secret material.

None

## Type of Change

- [x] 🐛 Bug fix (non-breaking change fixing an issue)
- [x] ✨ New feature (non-breaking change adding functionality)
- [ ] 💥 Breaking change (fix or feature causing existing functionality to change)
- [x] 📚 Documentation update
- [ ] 🏗️ Infrastructure change (Terraform/IaC)
- [ ] ♻️ Refactoring (no functional changes)

## Component(s) Affected

- [ ] `infrastructure/terraform/prerequisites/` - Azure subscription setup
- [ ] `infrastructure/terraform/` - Terraform infrastructure
- [x] `infrastructure/setup/` - OSMO control plane / Helm
- [ ] `workflows/` - Training and evaluation workflows
- [x] `training/` - Training pipelines and scripts
- [x] `docs/` - Documentation

## Testing Performed

- [ ] Terraform `plan` reviewed (no unexpected changes)
- [ ] Terraform `apply` tested in dev environment
- [ ] Training scripts tested locally with Isaac Sim
- [ ] OSMO workflow submitted successfully
- [ ] Smoke tests passed (`smoke_test_azure.py`)

None of the listed items apply: this PR changes no Terraform, submits no OSMO workflow, and touches no Isaac Sim path. Validation performed instead:

| Check | Result |
| --- | --- |
| `npm run lint:md` (tracked files) | 0 issues |
| `npm run spell-check` | 0 issues across 791 files |
| `npm run format:tables` | clean |
| `npm run lint:sh` | 95 files, 0 errors, 0 warnings |
| `npm run lint:py` | passed |
| `npm run lint:uvlock` (changed-files-only, as PR validation runs it) | passed |
| Frontend `npm run validate` | 1669/1669 tests, 0 errors |
| Backend `pytest` (changed modules) | 107 passed, 15 skipped |

Hardware validation on an NVIDIA Thor (aarch64) host, inside the rebuilt dev container as the non-root `vscode` user:

- `torch.cuda.is_available()` returned `True`, device reported as `NVIDIA Thor`, CUDA runtime 13.0, and a 2048x2048 matmul produced finite values.
- `get_safe_default_video_backend()` resolved to `pyav` with `torchcodec` absent, confirming the decoder is consistent across architectures.
- The `osmo` 6.3.0 arm64 client installed and reported its version.

Environment: local NVIDIA Thor workstation only. No Azure resources were deployed and no cost was incurred.

Known limitations: the x86_64 path was not executed on hardware, so the TorchCodec-to-PyAV decoder change is unverified for AzureML and OSMO training runs. LeRobot training itself was not run end-to-end on ARM64; validation covered CUDA availability, dependency resolution, and backend selection.

## Documentation Impact

- [ ] No documentation changes needed
- [x] Documentation updated in this PR
- [ ] Documentation issue filed

*docs/contributing/prerequisites.md* gains a Dev Container GPU Runtime section covering host runtime configuration for CSV-mode ARM64 hosts and the device-group access requirements, including how to confirm the host `render` GID and the misleading error-801 symptom. *docs/getting-started/quickstart.md* cross-links to it.

## Bug Fix Checklist

*Complete this section for bug fix PRs. Skip for other contribution types.*

- [x] Linked to issue being fixed — Closes #1446
- [x] Regression test included, OR
- [ ] Justification for no regression test:

Regression coverage added for each fix: two TrajectoryEditor tests (pending edits surviving an unrelated store update, and inputs clearing on Clear All), blob-sync tests asserting directory and `.cache` blobs are not downloaded, and a handler test proving an invalid cached clip is regenerated. The TrajectoryEditor tests were confirmed to fail against the pre-fix implementations.

## Checklist

- [x] My code follows the [project conventions](copilot-instructions.md)
- [x] Commit messages follow [conventional commit format](instructions/commit-message.instructions.md)
- [x] I have performed a self-review
- [x] Documentation impact assessed above
- [x] No new linting warnings introduced

## Notes

The 39 frontend lint warnings are pre-existing `react-hooks/set-state-in-effect` findings; this PR removes one such effect and adds none.

`npm run lint:uvlock` passes for all 13 locked Python projects after the review update.

Follow-up candidates, all evidenced in the diff:

- The `datasets==5.0.0` override comment in *training/il/lerobot/pyproject.toml* cites a cap from LeRobot's `dataset` extra, which is no longer selected; confirm whether the override is still required.

- The `render` GID `993` in *.devcontainer/devcontainer.json* is host-specific. It is documented in both the file comment and the prerequisites guide, but it is not resolved dynamically.

## Review amendment — September 1, 2026

Closes #1446

Reviewer feedback is addressed in `a43555e0`:

- Digest-pinned the final dev-container runtime image and removed the unrestricted base-image build argument so container discovery scans the shipped image.
- Enabled all NVIDIA driver capabilities, including Vulkan.
- Contained metadata-template and legacy video paths within the dataset root, rejecting absolute paths and traversal.
- Added real LeRobot shared-file integration coverage for metadata loading, Parquet episode filtering, path templates, timestamp-window video decoding, empty episodes, and legacy fallbacks.
- Added direct ffmpeg probe coverage for valid clips, failures, timeouts, and environments without ffmpeg.
- Added PyAV and PyArrow to the evaluation CI test group so integration coverage executes rather than skips.
- Isolated Dataviewer API tests from developer Azure storage configuration to prevent credential retries.

Validation after the review fixes: evaluation 300 passed with 96.25% coverage; Dataviewer backend 1,136 passed; Dataviewer frontend 1,669 passed; LeRobot download tests 67 passed; Python lint, uv lock consistency, and focused container-image discovery tests passed. Reviewer-provided AzureML and OSMO end-to-end runs also passed.

### CODE REVIEW

File: evaluation/sil/scripts/run-local-lerobot-eval.py [hunk 1]
```diff
@@ -69,7 +70,28 @@ def resolve_device(requested: str) -> str:
     return "cpu"
 
 
-def find_data_file(dataset_dir: str, ep_idx: int, info: dict) -> str | None:
+def load_episode_metadata(dataset_dir: str) -> dict[int, dict]:
```
Comment (katriendg) [inline evaluation/sil/scripts/run-local-lerobot-eval.py:73]: The new shared-file path is the main compatibility change, but its metadata read, path-template resolution, shared-parquet filtering, and empty-episode behavior are currently uncovered. Please add fixture-backed tests for these branches, including the legacy fallback. This is also the source of 13 missing and 6 partial patch lines in Codecov, could you please enhance code coverage tests here and in the changes covered below?

File: evaluation/sil/scripts/run-local-lerobot-eval.py [hunk 2]
```diff
@@ -82,7 +104,25 @@ def find_data_file(dataset_dir: str, ep_idx: int, info: dict) -> str | None:
     return None
 
 
-def find_video_file(dataset_dir: str, video_key: str, ep_idx: int, info: dict) -> str | None:
+def find_video_file(
+    dataset_dir: str,
```
Comment (katriendg) [inline evaluation/sil/scripts/run-local-lerobot-eval.py:123]: Please resolve metadata-derived data/video candidates against the dataset root and reject absolute paths or any result outside that root. The blob downloader already applies this containment pattern. Add tests for absolute paths and `..` traversal before passing dataset-controlled files to Parquet or native video decoders.

File: .devcontainer/Dockerfile [hunk 3]
```diff
@@ -1,5 +1,26 @@
 ARG BASE_IMAGE=
-FROM ${BASE_IMAGE:-mcr.microsoft.com/devcontainers/python:3.12-bookworm}
+FROM nvidia/cuda:13.0.3-cudnn-runtime-ubuntu24.04@sha256:14f6d08d1cd4a96effbfe3101d0b56326f552c199d05e4979ee0bd616df5811b AS cuda-runtime
```
Comment (katriendg) [inline .devcontainer/Dockerfile:1]: The CUDA stage is digest-pinned and the corresponding image scan passed. The final stage defaults to mutable `mcr.microsoft.com/devcontainers/base:ubuntu-24.04` and accepts an unrestricted build argument. CI's image discovery scanned the literal CUDA stage but did not report a scan for the parameterized MCR final base.

**Suggestion**

Please digest-pin the default MCR runtime base and ensure the effective final image is included in container scanning. The CUDA stage is pinned and scanned, but the shipped runtime still floats on a mutable tag, and the unrestricted `BASE_IMAGE` argument prevents CI from proving which final base was built.

File: data-management/viewer/backend/src/api/services/dataset_service/lerobot_handler.py [hunk 4]
```diff
@@ -495,6 +496,35 @@ def _video_cache_path(self, dataset_id: str, episode_idx: int, camera: str) -> P
         safe_camera = camera.replace("/", "_").replace("\\", "_")
         return loader.base_path / "meta" / "videos" / safe_camera / f"episode_{episode_idx:06d}.mp4"
 
+    @staticmethod
+    def _is_valid_video_file(video_path: Path) -> bool:
```
Comment (katriendg) [inline data-management/viewer/backend/src/api/services/dataset_service/lerobot_handler.py:500]: The new `_is_valid_video_file()` behavior is mocked by the regeneration test, leaving the ffmpeg probe itself uncovered. Could you add focused tests for a valid clip, an ffmpeg failure or timeout, and the no-ffmpeg fallback? This accounts for 8 missing and 1 partial changed lines in Codecov and would verify that valid cached clips are retained while corrupt ones are regenerated.

**Review by katriendg — COMMENTED** [review]
Thank you for the thorough contribution @IoTDan! The ARM64 support, LeRobot 0.6.1 migration, hardware validation, dependency pinning, and documentation are valuable and notably complete.

I left a few inline comments covering targeted testing and security concerns. Please also create and link the backlog issue for this work before we merge. Once those items are addressed, this should be ready for another review

For the issue, create **Feature** issue, then link the PR body with `Closes #<new-number>`.
Proposed issue title: `feat(training): support LeRobot 0.6.1 on NVIDIA ARM64`

**Comment by algattik** [comment]
We could consider adding an integration regression for  run-local-lerobot-eval.py  using real Parquet and video files in the LeRobot 0.6.1 shared-file layout. It should contain at least two episodes in one data file and verify  episode_index  filtering, path-template resolution and timestamp-window video decoding without mocking PyArrow or PyAV.

The accompanying tests add only a mocked timestamp-window case. They do not exercise the new metadata reader, real Arrow filtering or template-based shared-file layout.

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -92,10 +92,27 @@
...
+		"gpu": "optional"
+	},
+
 	// Container environment variables
 	"containerEnv": {
 		"BROWSER": "code --new-window",
-		"UV_LINK_MODE": "copy"
+		"UV_LINK_MODE": "copy",
+		"NVIDIA_VISIBLE_DEVICES": "all",
+		"NVIDIA_DRIVER_CAPABILITIES": "compute,utility"
```
Comment (agreaves-ms) [inline .devcontainer/devcontainer.json]: Can we set this to `"all"`, otherwise it looks like Vulkan driver capabilities may be omitted?

**Review by agreaves-ms — COMMENTED** [review]
This is looking great @IoTDan , after addressing the other comments then it looks like this PR is ready to go in. Thank you for the contribution!

**Comment by algattik** [comment]
I've run the following e2e tests against this PR, and they passed.

- test_e2e_aml_il_pipeline
- test_e2e_aml_il_lifecycle
- test_e2e_osmo_il_lifecycle

File: evaluation/sil/scripts/run-local-lerobot-eval.py (hunk 1 above)
Comment (IoTDan) [reply evaluation/sil/scripts/run-local-lerobot-eval.py:73]: Addressed in a43555e0. The regression now creates a real LeRobot 0.6.1 shared layout with two episodes in one Parquet/video file, invokes the production evaluation flow, verifies episode_index filtering, metadata path templates, timestamp-window video decoding, empty filtered episodes, and legacy fallback behavior.

File: evaluation/sil/scripts/run-local-lerobot-eval.py (hunk 2 above)
Comment (IoTDan) [reply evaluation/sil/scripts/run-local-lerobot-eval.py:123]: Addressed in a43555e0. Dataset-controlled paths now pass through a shared resolver that rejects absolute paths and resolved paths outside the dataset root. Tests cover metadata data/video templates, traversal attempts, and the legacy video-key fallback.

File: .devcontainer/Dockerfile (hunk 3 above)
Comment (IoTDan) [reply .devcontainer/Dockerfile:1]: Addressed in a43555e0. The final MCR base is now a literal digest-pinned FROM reference and the unrestricted BASE_IMAGE build argument was removed, allowing static image discovery to identify the shipped runtime. The focused image-discovery Pester suite passes (8 tests).

File: data-management/viewer/backend/src/api/services/dataset_service/lerobot_handler.py (hunk 4 above)
Comment (IoTDan) [reply data-management/viewer/backend/src/api/services/dataset_service/lerobot_handler.py:500]: Addressed in a43555e0 and 90523f9e. Coverage now verifies successful ffmpeg validation, nonzero exit, timeout, no-ffmpeg fallback, invalid-cache regeneration, and that a valid cached clip is retained without regeneration.

File: .devcontainer/devcontainer.json (hunk 5 above)
Comment (IoTDan) [reply .devcontainer/devcontainer.json]: Addressed in a43555e0. NVIDIA_DRIVER_CAPABILITIES is now set to all so the devcontainer exposes Vulkan in addition to compute and utility capabilities.

**Comment by IoTDan** [comment]
Review follow-up is complete. @algattik, the requested real shared-file integration regression now runs through the production evaluation flow and covers episode filtering, timestamp-window decoding, path templates, and empty episodes. @katriendg, Feature issue #1446 was created and linked with `Closes #1446`. @agreaves-ms, NVIDIA driver capabilities are now set to `all`. The remaining changed-line coverage gaps were closed in 90523f9e by testing Azure directory markers and valid cached-clip retention. Focused regressions and Python lint pass. The GHAS report remains informational; no new actionable dependency findings were reported.

**Comment by IoTDan** [comment]
All five inline threads have implementation replies. GitHub does not grant the PR author permission to resolve these reviewer-owned threads, so @katriendg and @agreaves-ms will need to mark their addressed threads resolved after verification.

**Review by katriendg — APPROVED** [review]
Thanks for addressing all feedback, I believe this looks good. Thanks a lot for this contribution!

**Review by agreaves-ms — APPROVED** [review]
